cms_coding_rules.md

```diff
@@ -0,0 +1,144 @@
+Coding rules are meant to prevent serious problems in software function, performance, maintainability, usability, and portability. The Packaging Rules section also has some brief guidelines for Python scripts.
+
+## 2 -- Naming Rules
+1. C++ header ﬁles use the sufﬁx .h, e.g. `CaloCluster.h` (*)
+2. For C++ source ﬁles, the preferred sufﬁx  is .cc, e.g. `CaloCluster.cc`, though there are existing files using .cpp and .cxx suffixes. (*)
+3. For a header file that contains a class, name that ﬁle after the class.
+4. Name source ﬁles after the class.
+5. For class, struct, type and enumeration names use upper class initials, e.g. `GeometryBuilder`.
+6. For namespaces use lower case, e.g. `namespace edm`.
+7. Start method names with lowercase, use upper case initials for following words, e.g. `collisionPoint()` (allowed exception: implementation of virtual methods inherited from external packages e.g. `ProcessHits()` method required by Geant4).
+8. Start data member names with lower case. A trailing “_” is the preferred method to distinguish a data member from the getter method (e.g. `momentum_`).
+9. Using “set” for a setter method is preferred, e.g. `setMomentum(double m)`.
+10. For a getter method, using the value name is preferred, e.g. `momentum()`.
+11. Do not use single character names, except for loop indices.
+12. Do not use special characters, except for “_” where allowed.
+13. Do not use “_” as ﬁrst character. Only use it as the last character for class data member names, not local variable names.
+14. Do not use “__”.
+15. Use clear and explanatory variable names.
+
+## 3 -- Style Rules
+1. Do not indent pre-processor directives -- there should be no leading spaces before a directive.  (*)
+2. Never change the language syntax using **#deﬁne**.
+3. Do not use spaces between method names and their argument list e.g. `foo()` rather than
+`foo ()`.
+4. Do not use spaces in front of [], () and on either side of -> . For example, `vector[i]` instead of `vector [i]`
+5. Separate expressions in a `for` statement by spaces.
+6. Use the same indentation for comments as for the block the comments refer to.
+
+## 4 -- Technical Coding Rules
+1. Protect each header ﬁle from multiple inclusion with  
+**#ifndef PackageName_SubPackageName_FileName_h  
+#define PackageName_SubPackageName_FileName_h**  
+(body of header file)  
+**#endif**  
+If necessary to create a unique name, one can add the directory name:  
+**PackageName_SubPackageName_Directory_FileName_h**.
+2. Each header ﬁle contains one class declaration only. (*)
+3. Header ﬁles must not contain any implementation except for class templates and code to be inlined.
+4. Do not inline virtual functions.
+5. Do not inline functions which contain control structures which require block scoping.
+6. In your own packages, use forward declarations if they are sufﬁcient.
+7. Do not forward-declare an entity from another package.
```

Comment:
Also add a new item:
If it will be used, creating a header file that only forward declares names and is located in the same package as the definitions of the names it declares is good practice.

Reply:
Again, in the interest of brevity, I don't think additional rules should be added. We are not trying to cover all cases, and we do reference the C++ Core Guidelines as a source for more detailed guidance.